generators/csharp/sdk/src/oauth/OauthTokenProviderGenerator.ts

```diff
@@ -0,0 +1,267 @@
+            directory: this.context.getCoreDirectory(),
+            allNamespaceSegments: this.context.getAllNamespaceSegments(),
+            allTypeClassReferences: this.context.getAllTypeClassReferences(),
+            namespace: this.context.getNamespace(),
+            customConfig: this.context.customConfig
+        });
+    }
+
+    protected getFilepath(): RelativeFilePath {
+        return join(this.context.getCoreDirectory(), RelativeFilePath.of(`${OAUTH_TOKEN_PROVIDER_CLASS_NAME}.cs`));
+    }
+
+    private getAccessTokenMethod(): csharp.Method {
+        return csharp.method({
+            access: "public",
+            isAsync: true,
+            name: "GetAccessTokenAsync",
+            parameters: [],
+            body: this.getAccessTokenBody(),
+            return_: csharp.Type.string()
+        });
+    }
+
+    private getAccessTokenBody(): csharp.CodeBlock {
+        const expiresIn = this.getExpiresIn();
+        const tokenEndpoint = this.scheme.configuration.tokenEndpoint;
+
+        return csharp.codeblock((writer) => {
+            writer.controlFlow(
+                "if",
+                csharp.codeblock((writer) => {
+                    writer.write(`${OauthTokenProviderGenerator.ACCESS_TOKEN_FIELD.name} == null`);
+                    // check expiresIn if present in the IR
+                    if (expiresIn != null) {
+                        writer.write(`|| DateTime.UtcNow >= ${OauthTokenProviderGenerator.EXPIRES_AT_FIELD.name}`);
+                    }
+                })
+            );
+
+            writer.writeNodeStatement(
+                csharp.codeblock((writer) => {
+                    const requestType = this.tokenEndpoint.requestBody?._visit({
+                        reference: (value) => {
+                            return this.context.csharpTypeMapper.convert({ reference: value.requestBodyType });
+                        },
+                        inlinedRequestBody: (value) => {
+                            return csharp.Type.reference(
+                                this.context.getRequestWrapperReference(
+                                    this.tokenEndpointReference.serviceId,
+                                    value.name
+                                )
+                            );
+                        },
+                        fileUpload: (value) => undefined,
+                        bytes: (value) => undefined,
+                        _other: (value) => undefined
+                    });
+                    if (requestType == null || requestType.internalType.type !== "reference") {
+                        throw new Error("Failed to get request class reference");
+                    }
+
+                    writer.write(`var tokenResponse = `);
+                    writer.writeNode(
+                        csharp.invokeMethod({
+                            async: true,
+                            method: `client.${this.tokenEndpoint.name.pascalCase.safeName}`,
```

Comment:
i should be able to call `snippetGenerator` here but that requires an IR Example, which is also impossible to construct at this point